storage/remote/queue_manager.go

```diff
@@ -503,41 +561,55 @@ func (s *shards) runShard(i int) {
 	}
 }
 
-func (s *shards) sendSamples(samples model.Samples) {
+func (s *shards) sendSamples(samples []tsdb.RefSample) {
 	begin := time.Now()
-	s.sendSamplesWithBackoff(samples)
+	err := s.sendSamplesWithBackoff(samples)
+	if err != nil {
+		level.Error(s.qm.logger).Log("msg", "non-recoverable error", "count", len(samples), "err", err)
+		return
+	}
 
 	// These counters are used to calculate the dynamic sharding, and as such
 	// should be maintained irrespective of success or failure.
 	s.qm.samplesOut.incr(int64(len(samples)))
 	s.qm.samplesOutDuration.incr(int64(time.Since(begin)))
+	if s.qm.storage != nil {
+		level.Debug(s.qm.logger).Log("pending", s.qm.storage.samplesIn.newEvents-s.qm.samplesOut.newEvents, "out", s.qm.samplesOut.newEvents, "in", s.qm.storage.samplesIn.newEvents)
+	}
 }
 
 // sendSamples to the remote storage with backoff for recoverable errors.
-func (s *shards) sendSamplesWithBackoff(samples model.Samples) {
+func (s *shards) sendSamplesWithBackoff(samples []tsdb.RefSample) error {
 	backoff := s.qm.cfg.MinBackoff
-	req := ToWriteRequest(samples)
-
-	for retries := s.qm.cfg.MaxRetries; retries > 0; retries-- {
-		begin := time.Now()
-		err := s.qm.client.Store(s.ctx, req)
-
-		sentBatchDuration.WithLabelValues(s.qm.queueName).Observe(time.Since(begin).Seconds())
-		if err == nil {
-			succeededSamplesTotal.WithLabelValues(s.qm.queueName).Add(float64(len(samples)))
-			return
-		}
+	s.qm.seriesMtx.Lock()
+	req := toWriteRequest(samples, s.qm.series)
+	s.qm.seriesMtx.Unlock()
+	for {
+		select {
+		case <-s.ctx.Done():
+			return nil
+		default:
+			begin := time.Now()
```

Comment:
Can we unnest this and just have an empty default?

Reply:
Just select on the ctx so it's non-blocking? Is there any benefit to not using the default case, or downside to using it?